onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -881,7 +887,27 @@ class PlannerImpl {
           if (!node_output->Exists()) continue;
           OrtValueIndex index = Index(node_output->Name());
           ProcessDef(index, node_output);
-          plan_.SetLocation(static_cast<size_t>(index), exec_provider->GetOrtDeviceByMemType(p_kernel_def->OutputMemoryType(i)));
+          OrtDevice output_device = exec_provider->GetOrtDeviceByMemType(p_kernel_def->OutputMemoryType(i));
+          // The below ifdef is due to GetConsumerNodes()
+          // Does minimal build support CPU based providers?
+#if !defined(ORT_MINIMAL_BUILD) || defined(ORT_EXTENDED_MINIMAL_BUILD)
+          if (output_device == OrtDevice()) {
+            const auto& output_name = node_output->Name();
+            const auto consumers = graph_viewer_.GetConsumerNodes(output_name);
+            for (const auto* consumer : consumers) {
+              if (consumer != nullptr) {
+                // Check if the provider is a compiling provider. They handle entire subgraphs at once with many inputs.
+                // Request input type on CPU, if not avaiable it would return the default.
+                const auto& ep_type = consumer->GetExecutionProviderType();
+                if (utils::ProviderIsCompiledType(ep_type)) {
+                  output_device = execution_providers_.Get(ep_type)->GetOrtDeviceByMemType(OrtMemType::OrtMemTypeCPUInput);
+                  break;
```

Comment:
Is the key aspect that the provider wants 4K alignment and not whether it compiles? 

If we had an alignment field in OrtDevice could be more generic? get the mem type for the default device, check if it's CPU memory, check if there's an alignment requirement.

Reply:
Yes, and no. Qnn may return either QtpSharedAllocator device or plain CPU, others currently would return CPU 4K aligned, or NPU allocator.

The key aspect here is to avoid the implicit copy, and the consumer EP knows what kind of allocator it needs depending on the build.

The alignment field alone would not do, because as I explain earlier, we need an OrtDevice instance that would uniquely id the allocator, even if we add alignment to the allocator implementation.

This concept could be extended to CPU -> CUDA allocations and avoid MemCpy node insertions. I.e. CUDAExecutionProvider may return OrtDevice with CUDAPinned memtype. This would cause the upstream output to be allocated in CPU accessible memory, so no copy would be necessary.